### PR TITLE
#3463 Master offers component 

### LINKED ISSUE #3463: Master Offers component v2
Great idea; I guess we can make a functional component with an address as param and input and merge these three current ones. 

When the logged-in user visits /bsx/offers it lands on its own address stats by default, i.e. in my case 
`/bsx/offers?u=bXgvP2cQzEnPvuPQHRghz9dUYovqzyPhXBRXbNwVCef7nsmqb`
where default would be incoming offers

There should be two filters where you can choose
- F1
  - created offers by particular address
  - incoming offers by particular address
- F2 
  - Active
  - Accepted
  - Withdrawn

When the non-logged user opens it, they should see a global table like at https://beta.kodadot.xyz/bsx/stats, so F1 should be disabled.

The goal would be for anyone/user can see per address offers 
- created
- incoming 
- global 

We can put it under the Stats dropdown in the navbar with the name Offers. 

Later it becomes the replacement for current stats

Work can start once the issues below are completed. 
- #3450 
- #3203 
- #2801

_Originally posted by @yangwao in https://github.com/kodadot/nft-gallery/issues/3462#issuecomment-1187044108_

### PR DESCRIPTION
**Pull request by prachi00**
**Thank you for your contribution** to the [KodaDot NFT gallery](https://kodadot.xyz).
👇 \_ Let's make a quick check before the contribution.

### PR type

- [ ] Bugfix
- [x] Feature
- [ ] Refactoring

### What's new?

- [x] PR closes #3463
- [ ] Requires deployment <rubick/magick_issue>
- [x] route is` /bsx/offers`

### Before submitting Pull Request, please make sure:

- [x] My contribution builds **clean without any errors or warnings**
- [x] I've merged recent default branch -- **main** and I've no conflicts
- [x] I've tried to respect high code quality standards
- [x] I've didn't break any original functionality
- [x] I've posted a screenshot of demonstrated change in this PR

### Optional

- [ ] I've tested it at </rmrk/collection/26902bc2f7c20c546a-1FVG7>
- [ ] I've tested PR on mobile and everything seems works
- [ ] I found edge cases
- [ ] I've written some unit tests 🧪

### Had issue bounty label?

- [x] Fill up your KSM address: 
[Payout](https://beta.kodadot.xyz/transfer/?target=EzGc4s9PgCPx1YnF3fqzhLzVHpHMTL4LWPScwpDrR8JKgSU)

### Community participation

- [x] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

### Screenshot

- [x] My fix has changed **something** on UI; a screenshot is best to understand changes for others.
Screen Shot 2022-08-06 at 5 17 53 PM

Screen Shot 2022-08-06 at 5 15 58 PM

### CODE REVIEW

**Comment by petersopko** [comment]
tested without wallet logged in, received 400 network error, pls check

**Review by petersopko — CHANGES_REQUESTED** [review]
error without wallet

**Comment by prachi00** [comment]
> https://deploy-preview-3692--koda-nuxt.netlify.app

where? because if you are not logged in, then the option is hidden 
Screen Shot 2022-08-08 at 1 04 47 PM

**Comment by petersopko** [comment]
So some feedback after checking this more thoroughly:

1. the tabs should be filter (F2) with two options as described in the issue, but let's make it three with first one being `All offers`, dropdown as the one for the offer status (F1). If there isn't valid address inserted, the F2 can be disabled dropdown (stuck at `All offers`), becoming enabled once there's valid address, giving two more filtering options, i.e. `Offers Created` and `Incoming Offers`

2. logged out user should be able to land at this page, and still see all of the offers (without entering address), data should be fetched the same way as in `/bsx/stats`

3. the entry field for address doesn't need `Check offers` button next to it, it could be 2 sec debounce and offers can be fetched once there's valid address

**Comment by prachi00** [comment]
> So some feedback after checking this more thoroughly:
> 
> 1. the tabs should be filter (F2) with two options as described in the issue, but let's make it three with first one being `All offers`, dropdown as the one for the offer status (F1). If there isn't valid address inserted, the F2 can be disabled dropdown (stuck at `All offers`), becoming enabled once there's valid address, giving two more filtering options, i.e. `Offers Created` and `Incoming Offers`
> 2. logged out user should be able to land at this page, and still see all of the offers (without entering address), data should be fetched the same way as in `/bsx/stats`
> 3. the entry field for address doesn't need `Check offers` button next to it, it could be 2 sec debounce and offers can be fetched once there's valid address

Regarding 2nd, can logged out users see the input box?

**Comment by prachi00** [comment]
Also, for all offers option, there should be no input box right?

**Comment by petersopko** [comment]
> can logged out users see the input box?

yes 

> also initially, when the user reloads, there will be address in the input box and option will be all, so wouldn't it be confusing that what the address id doing?

Hmm maybe we don't understand each other here, if there will be an address in the input field, you will be fetching all of the offers related to the address (incoming offers and made offers), if there isn't address, you're fetching all of the offers.

**Comment by prachi00** [comment]
> > can logged out users see the input box?
> 
> yes
> 
> > also initially, when the user reloads, there will be address in the input box and option will be all, so wouldn't it be confusing that what the address id doing?
> 
> Hmm maybe we don't understand each other here, if there will be an address in the input field, you will be fetching all of the offers related to the address (incoming offers and made offers), if there isn't address, you're fetching all of the offers.
Screen Shot 2022-08-10 at 11 01 57 AM
 okay, so logged in users will see these 3 options right and there is address in the input field by default for them, so which filter do they see first, `incoming` or `created`?

**Comment by prachi00** [comment]
> > > can logged out users see the input box?
> > 
> > 
> > yes
> > > also initially, when the user reloads, there will be address in the input box and option will be all, so wouldn't it be confusing that what the address id doing?
> > 
> > 
> > Hmm maybe we don't understand each other here, if there will be an address in the input field, you will be fetching all of the offers related to the address (incoming offers and made offers), if there isn't address, you're fetching all of the offers.
> 
> Screen Shot 2022-08-10 at 11 01 57 AM
> 
> okay, so logged in users will see these 3 options right and there is address in the input field by default for them, so which filter do they see first, `incoming` or `created`?



Also, if logged out users can see the input box, can they enter address and view incoming or created? because I was under the assumption that they cant do that

**Comment by prachi00** [comment]
> Also, if logged out users can see the input box, can they enter address and view incoming or created? because I was under the assumption that they cant do that

@petersopko ⬆️

**Comment by petersopko** [comment]
> Also, if logged out users can see the input box, can they enter address and view incoming or created? because I was under the assumption that they cant do that

yes, they can enter address and view incoming/created. input field is always enabled to enter an address, thing which is disabled is the filter until the valid address it's entered

**Comment by petersopko** [comment]
@prachi00 any updates on this one? :)

**Comment by prachi00** [comment]
> @prachi00 any updates on this one? :)

sorry about the delay, I was out the whole last week but getting back to this from today

**Comment by prachi00** [comment]
@petersopko initially when the user first opens it, should they see their own offers or general offers same as stats?

**Comment by petersopko** [comment]
> @petersopko initially when the user first opens it, should they see their own offers or general offers same as stats?

to quote from the initial issue description

> When the logged-in user visits /bsx/offers it lands on its own address stats by default, i.e. in my case
> /bsx/offers?u=bXgvP2cQzEnPvuPQHRghz9dUYovqzyPhXBRXbNwVCef7nsmqb
> where default would be incoming offers

**Comment by prachi00** [comment]
@petersopko fixed the issues mentioned

**Comment by petersopko** [comment]
> @petersopko fixed the issues mentioned

Test failed, and deepsource seems to have some kind of problem with the code tho ?

**Comment by petersopko** [comment]
On mobile it looks like this
image

**Comment by prachi00** [comment]
> On mobile it looks like this image

just merged with the main branch, should be fixed now

**Review by petersopko — CHANGES_REQUESTED** [review]
1. if I visit `/bsx/offers` without being logged in, and I log in, my address doesn't fill out
2. if I visit `/bsx/offers` while logged in, my address doesn't fill the search
3. the only time address input bar behaves correctly is when I get there through navbar, please fix that
4. please rename `Offer stats` to `Offers stats`

**Comment by petersopko** [comment]
I have fixed the failing tests for you, please check [this part of CONTRIBUTING.md](https://github.com/kodadot/nft-gallery/blob/main/CONTRIBUTING.md#keeping-tests-relevant) for further reference on how to make the tests happy once you change some of the tested subjects :)

**Comment by prachi00** [comment]
> I have fixed the failing tests for you, please check [this part of CONTRIBUTING.md](https://github.com/kodadot/nft-gallery/blob/main/CONTRIBUTING.md#keeping-tests-relevant) for further reference on how to make the tests happy once you change some of the tested subjects :)

got it, thank you for fixing this time, I'll make sure to do it from next time

**Comment by prachi00** [comment]
> 4\. Offers

for 1 and 2, I actually depend on the url to fill out the address in the input
`/bsx/offers?target=bXjGftu3aagU6RzMCKHi66X84jap3RKtvv6t1kxD2KXbJMiG2`

File: components/bsx/Offer/MasterOfferTable.vue [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+      } else {
+        showNotification('Unable to use target address', notificationTypes.warn)
+      }
+    } else {
+      this.skipUserOffer = true
+    }
+  }
+
+  public checkOfferForAddress(skipAddress?) {
+    this.skipUserOffer = skipAddress
```
Comment (kkukelka) [inline components/bsx/Offer/MasterOfferTable.vue:151]: the method is called once without giving any value to `skipAddress`so I'm not sure what will happen when setting `skipUserOffer` to optional value. I think its better to just set it to false as default value
```suggestion
  public checkOfferForAddress(skipAddress = false) {
    this.skipUserOffer = skipAddress
```

File: components/bsx/Offer/MyOffer.vue [hunk 2]
```diff
@@ -155,6 +167,12 @@ export default class MyOffer extends mixins(PrefixMixin, OfferMixin) {
   onAccountChange() {
     this.fetchMyOffers()
   }
+
+  @Watch('address', { immediate: true })
+  onAddressChange(value) {
```
Comment (kkukelka) [inline components/bsx/Offer/MyOffer.vue:172]:
```suggestion
  onAddressChange(value: string) {
```

**Review by kkukelka — CHANGES_REQUESTED** [review]

**Comment by prachi00** [comment]
> > 4. Offers
> 
> for 1 and 2, I actually depend on the url to fill out the address in the input `/bsx/offers?target=bXjGftu3aagU6RzMCKHi66X84jap3RKtvv6t1kxD2KXbJMiG2` because not doing this was creating a lot of edge cases for example, if you erase the input bar to view all the offers and then reload, it would show user's offers again

@petersopko let me know if this seems okay to you?

**Review by kkukelka — APPROVED** [review]
lg now

**Comment by yangwao** [comment]
let's resolve conflicts :)

**Comment by prachi00** [comment]
> let's resolve conflicts :)

done

**Review by petersopko — APPROVED** [review]
ok lgtm

**Comment by petersopko** [comment]
> @petersopko let me know if this seems okay to you?

it's not ideal, but this issue wasn't defined in great detail, so let's follow up with some tweaks

**Comment by petersopko** [comment]
pay 150 usd

**Comment by yangwao** [comment]
😍 Perfect, I’ve sent the payout 
💵 $150 @ 45.38 USD/KSM ~ 3.305 $KSM
🧗 [EzGc4s9PgCPx1YnF3fqzhLzVHpHMTL4LWPScwpDrR8JKgSU](https://kusama.subscan.io/account/EzGc4s9PgCPx1YnF3fqzhLzVHpHMTL4LWPScwpDrR8JKgSU)
🔗 [0x6ae7f9ab6b7fc3ea6a5dd18664c064f823e5b730ab15ec746513dc9f50c0c07b](https://kusama.subscan.io/extrinsic/0x6ae7f9ab6b7fc3ea6a5dd18664c064f823e5b730ab15ec746513dc9f50c0c07b)

🪅 Let’s grab another issue and get rewarded! 
🪄 [github.com/kodadot/nft-gallery/issues](https://github.com/kodadot/nft-gallery/issues/)

**Pull request merged**